src/transforms/kv_parser.rs

```diff
@@ -0,0 +1,404 @@
+    pub field_split: String,
+    #[derivative(Default(value = "true"))]
+    pub overwrite_target: bool,
+    pub separator: String,
+    pub target_field: Option<Atom>,
+    pub trim_key: Option<String>,
+    pub trim_value: Option<String>,
+    pub types: HashMap<Atom, String>,
+}
+
+inventory::submit! {
+    TransformDescription::new::<KeyValueConfig>("kv_parser")
+}
+
+#[typetag::serde(name = "kv_parser")]
+impl TransformConfig for KeyValueConfig {
+    fn build(&self, _cx: TransformContext) -> crate::Result<Box<dyn Transform>> {
+        let field = self
+            .field
+            .as_ref()
+            .unwrap_or(&event::log_schema().message_key());
+
+        let conversions = parse_conversion_map(&self.types)?;
+
+        Ok(Box::new(KeyValue {
+            conversions,
+            drop_field: self.drop_field,
+            field: field.clone(),
+            field_split: self.field_split.clone(),
+            overwrite_target: self.overwrite_target,
+            separator: self.separator.clone(),
+            target_field: self.target_field.clone(),
+            trim_key: self.trim_key.clone(),
+            trim_value: self.trim_value.clone(),
+        }))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "kv_parser"
+    }
+}
+
+pub struct KeyValue {
+    conversions: HashMap<Atom, Conversion>,
+    drop_field: bool,
+    field: Atom,
+    field_split: String,
+    overwrite_target: bool,
+    separator: String,
+    target_field: Option<Atom>,
+    trim_key: Option<String>,
+    trim_value: Option<String>,
+}
+
+fn kv_parser(
+    pair: String,
+    field_split: &String,
+    trim_key: &Option<String>,
+    trim_value: &Option<String>,
+) -> Option<(Atom, String)> {
+    let pair = pair.trim();
+    let mut field;
+
+    if field_split.is_empty() {
+        let mut kv_pair = pair.split_whitespace();
+        let key = kv_pair.nth(0)?;
+        let val = kv_pair.nth(1)?;
+        let count = kv_pair.count();
+
+        if count < 2 {
+            return None;
+        } else if count > 2 {
+            error!(
+                message = "KV parser expected 2 values, but got {count}",
+                count = count,
+                rate_limit_secs = 30
+            );
+            return None;
+        }
+
+        field = [key, val];
+    } else {
+        let split_index = pair.find(field_split).unwrap_or(0);
+        let (key, val) = pair.split_at(split_index);
+        let key = key.trim();
+        let val = val.trim_start_matches(field_split).trim();
+
+        if key.is_empty() || val.is_empty() {
+            return None;
+        }
+
+        field = [key, val];
+    }
+
+    for (i, trim) in [trim_key, trim_value].iter().enumerate() {
+        if let Some(trim) = trim {
+            let trim_chars: Vec<char> = trim.chars().collect();
```

Comment:
Since this will be constant, it should be computed when `struct KeyValue` is created rather than once for each line.

Reply:
Yup, great catch, that should increase perf and reduce memory usage a bit. I'll address this when I refactor to move the `kv_parser` function to a method.